src/worker/activities/activity_heartbeat_manager.rs

```diff
@@ -105,143 +111,221 @@ impl ActivityHeartbeatManager {
 }
 
 #[derive(Debug)]
-struct HbStreamState {
-    last_sent: HashMap<TaskToken, ActivityHbState>,
-    incoming_hbs: UnboundedReceiver<HBAction>,
+struct ActivityHeartbeatState {
+    task_token: TaskToken,
+    last_recorded_details: Option<Vec<common::Payload>>,
+    last_send_requested: Instant,
+    throttle_interval: Duration,
+    sleep_cancel_tx: Option<CancellationToken>,
 }
 
-impl HbStreamState {
-    fn new(incoming_hbs: UnboundedReceiver<HBAction>) -> Self {
-        Self {
-            last_sent: Default::default(),
-            incoming_hbs,
-        }
+impl ActivityHeartbeatState {
+    /// Get duration to sleep by subtracting `throttle_interval` by elapsed time since
+    /// `last_send_requested`
+    fn get_throttle_sleep_duration(&self) -> Duration {
+        let time_since_last_sent = self.last_send_requested.elapsed();
+
+        return if time_since_last_sent > Duration::ZERO
+            && self.throttle_interval > time_since_last_sent
+        {
+            self.throttle_interval - time_since_last_sent
+        } else {
+            Duration::ZERO
+        };
     }
 }
 
 #[derive(Debug)]
-struct ActivityHbState {
-    delay: Duration,
-    last_sent: Instant,
+struct HeartbeatStreamState {
+    tt_to_state: HashMap<TaskToken, ActivityHeartbeatState>,
+    incoming_hbs: UnboundedReceiver<HeartbeatAction>,
+    /// Token that can be used to cancel the entire stream.
+    /// Requests to the server are not cancelled with this token.
+    pub cancellation_token: CancellationToken,
+}
+
+impl HeartbeatStreamState {
+    fn new() -> (Self, UnboundedSender<HeartbeatAction>) {
+        let (heartbeat_tx, incoming_hbs) = unbounded_channel();
+        (
+            Self {
+                cancellation_token: Default::default(),
+                tt_to_state: Default::default(),
+                incoming_hbs,
+            },
+            heartbeat_tx,
+        )
+    }
+
+    /// Record a heartbeat received from lang
+    fn record(&mut self, hb: ValidActivityHeartbeat) -> Option<HeartbeatExecutorAction> {
+        match self.tt_to_state.entry(hb.task_token.clone()) {
+            Entry::Vacant(e) => {
+                let state = ActivityHeartbeatState {
+                    task_token: hb.task_token.clone(),
+                    throttle_interval: hb.throttle_interval,
+                    last_send_requested: Instant::now(),
+                    last_recorded_details: None,
```

Comment:
Shouldn't this be set here since it is in the other branch? If not seemingly it should be set in `complete_report` rather than the other branch

Reply:
We don't record it here because we send the details out.
We use `None` to denote that after throttling we can stop tracking this task token.
Adding a comment.